src/natcap/invest/spec.py

```diff
@@ -0,0 +1,2227 @@
+        """
+        file_warning = super().validate(filepath)
+        if file_warning:
+            return file_warning
+
+        try:
+            gdal_dataset = gdal.OpenEx(filepath, gdal.OF_VECTOR)
+        except RuntimeError:
+            return get_message('NOT_GDAL_VECTOR')
+
+        geom_map = {
+            'POINT': [ogr.wkbPoint, ogr.wkbPointM, ogr.wkbPointZM,
+                      ogr.wkbPoint25D],
+            'LINESTRING': [ogr.wkbLineString, ogr.wkbLineStringM,
+                           ogr.wkbLineStringZM, ogr.wkbLineString25D],
+            'POLYGON': [ogr.wkbPolygon, ogr.wkbPolygonM,
+                        ogr.wkbPolygonZM, ogr.wkbPolygon25D],
+            'MULTIPOINT': [ogr.wkbMultiPoint, ogr.wkbMultiPointM,
+                           ogr.wkbMultiPointZM, ogr.wkbMultiPoint25D],
+            'MULTILINESTRING': [ogr.wkbMultiLineString, ogr.wkbMultiLineStringM,
+                                ogr.wkbMultiLineStringZM,
+                                ogr.wkbMultiLineString25D],
+            'MULTIPOLYGON': [ogr.wkbMultiPolygon, ogr.wkbMultiPolygonM,
+                             ogr.wkbMultiPolygonZM, ogr.wkbMultiPolygon25D]
+        }
+
+        allowed_geom_types = []
+        for geom in self.geometry_types:
+            allowed_geom_types += geom_map[geom]
+
+        # NOTE: this only checks the layer geometry type, not the types of the
+        # actual geometries (layer.GetGeometryTypes()). This is probably equivalent
+        # in most cases, and it's more efficient than checking every geometry, but
+        # we might need to change this in the future if it becomes a problem.
+        # Currently not supporting ogr.wkbUnknown which allows mixed types.
+        layer = gdal_dataset.GetLayer()
+        if layer.GetGeomType() not in allowed_geom_types:
+            return get_message('WRONG_GEOM_TYPE').format(allowed=self.geometry_types)
+
+        if self.fields:
+            field_patterns = []
+            for spec in self.fields:
+                # brackets are a special character for our args spec syntax
+                # they surround the part of the key that's user-defined
+                # user-defined rows/columns/fields are not validated here, so skip
+                if spec.required is True and '[' not in spec.id:
+                    field_patterns.append(spec.id)
+
+            fieldnames = [defn.GetName() for defn in layer.schema]
+            required_field_warning = check_headers(
+                field_patterns, fieldnames, 'field')
+            if required_field_warning:
+                return required_field_warning
+
+        srs = layer.GetSpatialRef()
+        projection_warning = _check_projection(srs, self.projected, self.projection_units)
+        return projection_warning
+
+
+@dataclasses.dataclass
+class RasterOrVectorInput(FileInput):
+    """An invest model input that can be either a single-band raster or a vector."""
```

Comment:
I think this gets us under 80 chars.

```suggestion
    """An invest model input of either a single-band raster or a vector."""
```